File: code/mincostmaxflow.py
'''
Solves the min-cost-max-flow problem. This is finding a flow
of maximal capacity (or of capacity at most maxf) with a 
minimal cost. Each edge has a capacity and a cost.

Time Complexity: O(min(N^2*M^2, N*M*F))
Space Complexity: O(N^2)

This solution is about 2 times slower than java.
'''

#edge = [to, cap, cost, rev, f]
INF = 10**15

def createGraph(n):
    return [[] for _ in range(n)]

def addEdge(graph, fr, to, cap, cost):
    graph[fr].append([to,cap,cost,len(graph[to]),0])
    graph[to].append([fr,0,-cost,len(graph[fr])-1,0])

#edge = [to, cap, cost, rev, f]
def bellmanFord(s):
    n = len(graph)
    for i in range(n): dist[i] = INF
    dist[s] = 0
    inqueue = [False]*n
    curflow[s] = INF
    q = [0]*n
    qt = 0
    q[qt] = s
    qt += 1
    qh = 0
    while (qh-qt)%n != 0:
        u = q[qh%n]
        inqueue[u] = False
        for i in range(len(graph[u])):
            e = graph[u][i]
            if(e[4] >= e[1]): continue
            v = e[0]
            ndist = dist[u] + e[2]
            if dist[v] > ndist:
                dist[v] = ndist
                prevnode[v] = u
                prevedge[v] = i
                curflow[v] = min(curflow[u], e[1]-e[4])
                if not inqueue[v]:
                    inqueue[v] = True
                    q[qt%n] = v
                    qt += 1
        qh += 1

#edge = [to, cap, cost, rev, f]
def minCostFlow(s, t, maxf):
    n = len(graph)

    flow = 0
    flowCost = 0
    while flow < maxf:
        bellmanFord(s)
        if dist[t] == INF: break
        df = min(curflow[t], maxf - flow)
        flow += df
        v = t
        while v != s:
            e = graph[prevnode[v]][prevedge[v]]
            graph[prevnode[v]][prevedge[v]][4] += df
            graph[v][e[3]][4] -= df
            flowCost += df*e[2]
            v = prevnode[v]
    return (flow, flowCost)

#Example of useage. MUST USE THE SAME NAMES!
N,M,S,T = map(int, raw_input().split())
graph = createGraph(N)
for i in range(M):
    U,V,C,W = map(int, raw_input().split())
    addEdge(graph, U, V, C, W)

dist = [INF]*N
curflow = [0]*N
prevedge = [0]*N
prevnode = [0]*N
flow, flowCost = minCostFlow(S, T, INF)
print flow, flowCost

File: code/mincostmaxflow.java
import java.util.*;
import java.io.*;
//Use just like python, same solution.
public class mincostmaxflownew {

  static class Edge {
    int to, f, cap, cost, rev;

    Edge(int v, int cap, int cost, int rev) {
      this.to = v;
      this.cap = cap;
      this.cost = cost;
      this.rev = rev;
    }
  }

  public static List<Edge>[] createGraph(int n) {
    List<Edge>[] graph = new List[n];
    for (int i = 0; i < n; i++)
      graph[i] = new ArrayList<Edge>();
    return graph;
  }

  public static void addEdge(List<Edge>[] graph, int s, int t, 
          int cap, int cost) {
    graph[s].add(new Edge(t, cap, cost, graph[t].size()));
    graph[t].add(new Edge(s, 0, -cost, graph[s].size() - 1));
  }

  static void bellmanFord(List<Edge>[] graph, int s, int[] dist, 
          int[] prevnode, int[] prevedge, int[] curflow) {
    int n = graph.length;
    Arrays.fill(dist, 0, n, Integer.MAX_VALUE);
    dist[s] = 0;
    curflow[s] = Integer.MAX_VALUE;
    boolean[] inqueue = new boolean[n];
    int[] q = new int[n];
    int qt = 0;
    q[qt++] = s;
    for (int qh = 0; (qh - qt) % n != 0; qh++) {
      int u = q[qh % n];
      inqueue[u] = false;
      for (int i = 0; i < graph[u].size(); i++) {
        Edge e = graph[u].get(i);
        if (e.f >= e.cap)
          continue;
        int v = e.to;
        int ndist = dist[u] + e.cost;
        if (dist[v] > ndist) {
          dist[v] = ndist;
          prevnode[v] = u;
          prevedge[v] = i;
          curflow[v] = Math.min(curflow[u], e.cap - e.f);
          if (!inqueue[v]) {
            inqueue[v] = true;
            q[qt++ % n] = v;
          }
        }
      }
    }
  }

  public static int[] minCostFlow(List<Edge>[] graph, int s, 
          int t, int maxf) {
    int n = graph.length;
    int[] dist = new int[n];
    int[] curflow = new int[n];
    int[] prevedge = new int[n];
    int[] prevnode = new int[n];

    int flow = 0;
    int flowCost = 0;
    while (flow < maxf) {
      bellmanFord(graph, s, dist, prevnode, prevedge, curflow);
      if (dist[t] == Integer.MAX_VALUE)
        break;
      int df = Math.min(curflow[t], maxf - flow);
      flow += df;
      for (int v = t; v != s; v = prevnode[v]) {
        Edge e = graph[prevnode[v]].get(prevedge[v]);
        e.f += df;
        graph[v].get(e.rev).f -= df;
        flowCost += df * e.cost;
      }
    }
    return new int[]{flow, flowCost};
  }

  // Usage example
  public static void main(String[] args) throws IOException{
    BufferedReader br = new BufferedReader(
            new InputStreamReader(System.in));
    String[] s = br.readLine().split(" ");
    int N = Integer.parseInt(s[0]);
    int M = Integer.parseInt(s[1]);
    int S = Integer.parseInt(s[2]);
    int T = Integer.parseInt(s[3]);
    List<Edge>[] graph = createGraph(N);
    for(int i = 0; i < M; i++){
        s = br.readLine().split(" ");
        int U = Integer.parseInt(s[0]);
        int V = Integer.parseInt(s[1]);
        int C = Integer.parseInt(s[2]);
        int W = Integer.parseInt(s[3]);
        addEdge(graph,U,V,C,W);
    }
    int[] res = minCostFlow(graph, S, T, Integer.MAX_VALUE);
    int flow = res[0];
    int flowCost = res[1];
    System.out.println(flow + " " + flowCost);
  }
}
